infra/ansible/files/register-worker.cgi

```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+echo "Content-type: text/html"
+echo ""
+
+if [ "$REQUEST_METHOD" != "POST" ]; then
+  echo 'Post only'
+  exit 0
+fi
+
+WORKER=$(cat - | sed -n 's/worker=//p')
+
+if [ -z "$WORKER" ]; then
+  echo 'Need worker=<hostname> in body'
+  exit 0
+fi
+
+echo docker node update --label-add TGRole=worker $WORKER
+docker node update --label-add TGRole=worker $WORKER
+
+exit 0
+~
```

Comment:
Are we installing an Apache Server, just so that we can run a single .cgi script to call `docker node update`? Seems a bit over the top. Could this be done by a tiny go program so that in the future is part of the testground daemon?

Reply:
I agree with @daviddias on this. In my afternoons worth of digging, we are already controlling docker swarm from within `pkg/runner/cluster_swarm.go` so it doesn't make sense to me that we are doing metadata coordination outside of this in the infrastructure layer, i.e. ansible, as well.

Reply:
@jimpick Using something like [AWS Launch Templates](https://capstonec.com/auto-scaling-docker-nodes-in-aws/) we could put a message onto redis and have the `manager` node listen for that?

Reply:
AFAICT `cluster_swarm.go` is used to run a `docker service` on an already established `docker swarm` cluster.

I believe we are applying labels to the workers here, so that we can assign workloads only to the workers through the use of labels... is that right @jimpick ? I think this actually should be done when we setup the infrastructure on the Terraform/Ansible layer, and not when we run `testground`.

I also think we should be able to do this without Apache web server though - we could just run `docker node update` as part of the Ansible playbook to setup the necessary labels on workers.

Reply:
The Apache server also serves the filebeat.yml, authorized_hosts and docker swarm join key to any new worker nodes that boot up.

I initially entertained the idea of writing a single-purpose HTTP server in Go that does the exact same thing as what Ansible does with installing Apache + the CGI script, but since we were already using Ansible, it seemed cleaner to me to keep it all together in one place.

Reply:
> AFAICT `cluster_swarm.go` is used to run a `docker service` on an already established `docker swarm` cluster.
> 
> I believe we are applying labels to the workers here, so that we can assign workloads only to the workers through the use of labels... is that right @jimpick ? I think this actually should be done when we setup the infrastructure on the Terraform/Ansible layer, and not when we run `testground`.
> 
> I also think we should be able to do this without Apache web server though - we could just run `docker node update` as part of the Ansible playbook to setup the necessary labels on workers.

Agree, this is a step that happens when preparing infrastructure before the `testground` binary is run. (I imagine the scope of the binary could be expanded to include managing infrastructure, but it currently doesn't do that)

This PR is for a boot script that gets run automatically on a new worker node when it is added to the swarm. It can't run Ansible commands on the manager node in order to run `docker node update` (that command can't be run on worker nodes), so that's why I had to do the CGI script. 

Alternatively, I could have distributed the private ssh key and installed it on the worker nodes (via HTTP) so they could run Ansible commands on the manager node, but that felt wrong to me.

Reply:
> @jimpick Using something like [AWS Launch Templates](https://capstonec.com/auto-scaling-docker-nodes-in-aws/) we could put a message onto redis and have the `manager` node listen for that?

Looks like they are just putting the Docker Swarm join key into the user data, which I think is possible with either Launch Templates or Launch Configurations. We could potentially do that because we create the Launch Configuration from Terraform, but I'd have to experiment ... it would have to create the Docker Swarm first, and then export the join key to Terraform somehow so it could use it when generating the Launch Configuration or Launch Template.

We do spin up Redis as a Docker service, but I don't think it's exposed externally.